smolyay/grid.py

```diff
@@ -0,0 +1,396 @@
+        grid points' functions = [(:class:`ChebyshevFunction`,
+        :class:`ChebyshevFunction`)
+        , (:class:`ChebyshevFirstKind`, :class:`ChebyshevFirstKind`),
+        (:class:`ChebyshevFirstKind`, :class:`ChebyshevFirstKind`),
+        (:class:`ChebyshevFirstKind`, :class:`ChebyshevFirstKind`),
+        (:class:`ChebyshevFirstKind`, :class:`ChebyshevFirstKind`)]
+
+    """
+
+    def __init__(self, basis_set):
+        super().__init__(basis_set)
+        if not isinstance(self.basis_set, NestedBasisFunctionSet):
+            raise TypeError('Basis set must be nested for Smolyak grid')
+
+    def __call__(self, dimension):
+        """Make grid points and their corresponding basis functions.
+
+        Depending on the dimensionality, a set of grid points are generated
+        (integer grid points). Grid points of the basis function, and
+        their corresponding basis function and levels are then
+        generated by replacing the indexes with their
+        corresponding unidimensional points, functions, and levels' indexes.
+
+        Parameters
+        ----------
+        dimension: int
+            Number of independent variables.
+
+        Returns
+        -------
+        :class:`NestedIndexGrid`
+            Grid points (and their indexes),levels,
+            and their corresponding functions.
+        """
+        grid_points_indexes = None
+        for sum_of_levels in range(dimension,
+                                   dimension+len(self._basis_set.levels)):
+            for composition in generate_compositions(
+                    sum_of_levels, dimension, include_zero=False):
+                index_composition = numpy.array(composition) - 1
+                # generate all combinations of
+                # the arrays along each dimension
+                level_composition_index = [self._basis_set.levels[index]
+                                           for index in index_composition]
+                grid_points_indexes_ = (numpy.array(
+                    numpy.meshgrid(*level_composition_index))
+                    .T.reshape(-1, dimension))
+                if grid_points_indexes is None:
+                    grid_points_indexes = grid_points_indexes_
+                else:
+                    grid_points_indexes = numpy.concatenate(
+                        (grid_points_indexes,
+                         grid_points_indexes_), axis=0)
+        # make integer grids, levels, grid points and basis functions
+        grid_points_indexes = grid_points_indexes.tolist()
+        level_lookup = {point: level for level, points in
+                        enumerate(self._basis_set.levels) for point in points}
+        levels = numpy.array([level_lookup[index] for index in
+                              numpy.concatenate(grid_points_indexes)]
+                             ).reshape((len(grid_points_indexes),
+                                        dimension)).tolist()
```

Comment:
Is this casting and reshaping necessary, or could it be done like:
```
levels = [[level_lookup[index] for index in indexes] for indexes in grid_point_indexes]
```
?

Reply:
It could be done. I was just trying to avoid having two for loops.